### PR TITLE
Implement GetZoneByProviderID & GetZoneByNodeName

### LINKED ISSUE #50926: Implement GetZoneByProviderID and GetZoneByNodeName for all cloud providers
**Is this a BUG REPORT or FEATURE REQUEST?**:
> /kind feature

As part of the initiative for external cloud providers (https://github.com/kubernetes/kubernetes/issues/48690), we added GetZoneByProviderID and GetZoneByNodeName. See https://github.com/kubernetes/kubernetes/pull/50858 for more details. The previous PR did not actually implement those methods for existing cloud providers in `pkg/cloudprovider/providers`. These methods should be implemented by all cloud providers in the future. 

cc @wlan0

### PR DESCRIPTION
**Pull request by cheftako**
Adding an implementation of GetZoneByProviderID & GetZoneByNodeName for
GCE.
This is related to ticket 50926.
This was tested as part of the ongoing separate GCE cloud provider work.

**What this PR does / why we need it**: It implements GCE methods needed by the cloud provider work.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #50926 

**Special notes for your reviewer**: Tested with pull/50811

**Release note**:
<!--  Steps to write your release note:
```release-note NONE
```

### CODE REVIEW

**Review by luxas — APPROVED** [review]
/lgtm

@wojtek-t looks good to you as well?

File: pkg/cloudprovider/providers/gce/gce_zones.go [hunk 1]
```diff
@@ -43,14 +43,34 @@ func (gce *GCECloud) GetZone() (cloudprovider.Zone, error) {
...
+	}
+	return cloudprovider.Zone{FailureDomain: zone, Region: region}, nil
 }
 
 // GetZoneByNodeName implements Zones.GetZoneByNodeName
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
 func (gce *GCECloud) GetZoneByNodeName(nodeName types.NodeName) (cloudprovider.Zone, error) {
-	return cloudprovider.Zone{}, errors.New("GetZoneByNodeName not imeplemented")
+	list, err := gce.service.Zones.List(gce.projectID).Filter(fmt.Sprintf("name eq %s", mapNodeNameToInstanceName(nodeName))).Do()
```
Comment (wojtek-t) [inline pkg/cloudprovider/providers/gce/gce_zones.go]: I'm not convinced about this. According to this documentation:
https://godoc.org/google.golang.org/api/compute/v1#ZonesService.List

It seems that Zones.List is listing zones.
So how passing instanceName as a filter for the name is supposed to work here?

File: pkg/cloudprovider/providers/gce/gce_test.go [hunk 2]
```diff
@@ -281,7 +282,7 @@ func TestSplitProviderID(t *testing.T) {
 	for _, test := range providers {
 		project, zone, instance, err := splitProviderID(test.providerID)
 		if (err != nil) != test.fail {
-			t.Errorf("Expected to failt=%t, with pattern %v", test.fail, test)
+			t.Errorf("Expected to fail=%t, with pattern %v", test.fail, test)
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/gce/gce_test.go]: %t prints the type, which is probably not very useful.

File: pkg/cloudprovider/providers/gce/gce_test.go [hunk 3]
```diff
@@ -300,6 +301,61 @@ func TestSplitProviderID(t *testing.T) {
...
+	}
+
+	gce := &GCECloud{
+		localZone: "us-central1-f",
+		region:    "us-central1",
+	}
+	for _, test := range tests {
+		zone, err := gce.GetZoneByProviderID(test.providerID)
+		if (err != nil) != test.fail {
+			t.Errorf("Expected to fail=%t, provider ID %v, tests %s", test.fail, test, test.description)
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/gce/gce_test.go]: %t probably wrong here, too

**Review by lavalamp — COMMENTED** [review]
only nits.

**Comment by lavalamp** [comment]
Oops, I guess %t is boolean; %T is type. TIL!

/lgtm
/approve

**Comment by wojtek-t** [comment]
LGTM - now it works.

**Pull request merged**